File: tiro.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <sys/time.h>
#include <unistd.h>

double distancia;


// Función que genera el código HTML para visualizar coordenadas en un mapa de Google
void generarCodigoHTML(float coordenadas[][2], int numCoordenadas, const char* apiKey) {


	float X = (coordenadas[0][0] + coordenadas[numCoordenadas - 1][0]) / 2;
	float Y = (coordenadas[0][1] + coordenadas[numCoordenadas - 1][1]) / 2;

    printf("<!DOCTYPE html>\n");
    printf("<html>\n");
    printf("  <head>\n");
    printf("    <title>Mapa con Coordenadas</title>\n");
    printf("    <script src=\"https://maps.googleapis.com/maps/api/js?key=%s\"></script>\n", apiKey);
    printf("  </head>\n");
    printf("  <body>\n");
    printf("    <div id=\"map\" style=\"height: 800px; width: 100%%;\"></div>\n");
    printf("    <script>\n");
    printf("      function initMap() {\n");
    printf("        var map = new google.maps.Map(document.getElementById(\"map\"), {\n");
    printf("          zoom: 18,\n");
    printf("          center: { lat: %.6f, lng: %.6f },\n",X,Y);
    printf("          mapTypeId: google.maps.MapTypeId.SATELLITE\n");
    printf("        });\n");

    // Recorrer las coordenadas y agregar marcadores
    for (int i = 0; i < numCoordenadas; i++) {
        printf("        new google.maps.Marker({\n");
        printf("          position: { lat: %.6f, lng: %.6f },\n", coordenadas[i][0], coordenadas[i][1]);
        printf("          map: map,\n");
        if (i==0) {
			printf("          label: \"Mtr\"\n");
        }else{
			if (i!=(numCoordenadas-1)) printf("          label: \"%d\"\n", i);
			else printf("          label: \"Obj\"\n");
		}
        //printf("          icon: \"http://maps.google.com/mapfiles/ms/micons/blue.png\"\n");
        printf("        });\n");
    }


// Agregar una línea con forma de flecha entre el primer y último punto con texto en el centro
    printf("        var linea = new google.maps.Polyline({\n");
    printf("          path: [\n");
    printf("            { lat: %.6f, lng: %.6f },\n", coordenadas[0][0], coordenadas[0][1]); // Primer punto
    printf("            { lat: %.6f, lng: %.6f }  // Último punto\n", coordenadas[numCoordenadas - 1][0], coordenadas[numCoordenadas - 1][1]);
    printf("          ],\n");
    printf("          geodesic: true,\n");
    printf("          icons: [{\n");
    printf("            icon: { path: google.maps.SymbolPath.FORWARD_CLOSED_ARROW },\n");
    printf("            offset: '100',\n");
    printf("          }],\n");
    printf("          strokeColor: '#FF0000', // Color de la línea (rojo)\n");
    printf("          strokeOpacity: 1.0, // Opacidad de la línea (1.0 es completamente opaco)\n");
    printf("          strokeWeight: 2 // Grosor de la línea\n");
    printf("        });\n");


    printf("        linea.setMap(map);\n"); // Agregar la línea al mapa

 // Crear un texto en el centro de la línea
    printf("        var centerLatLng = new google.maps.LatLng(%.6f, %.6f); \n",X,Y);
    printf("        var textoEnCentro = new google.maps.InfoWindow({\n");
    printf("          content: '%.2f m',\n",distancia);
    printf("        });\n");
    printf("        textoEnCentro.setPosition(centerLatLng);\n");
    printf("        textoEnCentro.open(map);\n");



    printf("      }\n");
    printf("    </script>\n");
    printf("    <script>\n");
    printf("      initMap();\n");
    printf("    </script>\n");
    printf("  </body>\n");
    printf("</html>\n");
}


void L2G(double decimal, int *grados, int *minutos, double *segundos) {
    // Verificar si el valor es negativo
    int signo = (decimal < 0) ? -1 : 1;
    decimal = fabs(decimal);

    // Calcular los grados
    *grados = (int)decimal;

    // Calcular los minutos
    double minutosDecimales = (decimal - *grados) * 60;
    *minutos = (int)minutosDecimales;

    // Calcular los segundos
    *segundos = (minutosDecimales - *minutos) * 60;

    // Aplicar el signo correcto
    *grados *= signo;
}

float addwobble(float valorInicial, float porcentajeMaximo) {

    // Generar un número aleatorio entre -1 y 1
    float signoAleatorio = ((float)rand() / RAND_MAX) * 2 - 1;

    // Calcular el valor aleatorio dentro del rango
    float valorAleatorio = signoAleatorio * porcentajeMaximo * valorInicial;

    // Calcular el valor final
    float valorFinal = valorInicial + valorAleatorio;

    return valorFinal;
}

int gA, mA, gO, mO;
double sA, sO;





// Función para calcular la distancia recorrida por un proyectil en una trayectoria parabólica
double calcularDistanciaProyectil(double angulo, double velocidadInicial) {
    const double gravedad = 9.81; // Asumimos una aceleración debida a la gravedad de 9.81 m/s^2

    // Convertir el ángulo de grados a radianes
    double anguloRadianes = angulo * M_PI / 180.0;

    // Calcular la distancia
    double distancia = (velocidadInicial * velocidadInicial * sin(2 * anguloRadianes)) / gravedad;

    return distancia;
}

// Función para calcular las coordenadas de impacto
void calcularCoordenadasImpacto(double latitudInicial, double longitudInicial, double direccion, double distancia, double *latitudImpacto, double *longitudImpacto) {
    // Radio de la Tierra en metros
    const double radioTierra = 6371000.0;

    // Convertir la dirección en grados a radianes
    double direccionRadianes = direccion * M_PI / 180.0;

    // Convertir latitud y longitud inicial a radianes
    double latitudRadianes = latitudInicial * M_PI / 180.0;
    double longitudRadianes = longitudInicial * M_PI / 180.0;

    // Calcular las coordenadas cartesianas del impacto
    double xImpacto = distancia * cos(direccionRadianes);
    double yImpacto = distancia * sin(direccionRadianes);

    // Calcular las coordenadas esféricas del impacto
    double nuevaLatitudRadianes = asin(sin(latitudRadianes) * cos(distancia / radioTierra) + cos(latitudRadianes) * sin(distancia / radioTierra) * cos(direccionRadianes));
    double nuevaLongitudRadianes = longitudRadianes + atan2(sin(direccionRadianes) * sin(distancia / radioTierra) * cos(latitudRadianes), cos(distancia / radioTierra) - sin(latitudRadianes) * sin(nuevaLatitudRadianes));

    // Convertir las coordenadas esféricas del impacto a grados
    *latitudImpacto = nuevaLatitudRadianes * 180.0 / M_PI;
    *longitudImpacto = nuevaLongitudRadianes * 180.0 / M_PI;
}


int main(int argc, char *argv[]) {

	// Obtener la marca de tiempo actual en milisegundos
    struct timeval tiempo;
    gettimeofday(&tiempo, NULL);
    unsigned long semilla = tiempo.tv_sec * 1000 + tiempo.tv_usec / 1000;

    srand((unsigned)semilla*(unsigned long)getpid());



    double velocidadInicial=90;

    if (argc != 7) {
        printf("Uso: %s <Lat> <Long> <disparos> <azimut> <direccion> <dispersion>\n", argv[0]);
        return 1;
    }
    double latitudInicial = atof(argv[1]);
    double longitudInicial = atof(argv[2]);
	int N = atoi(argv[3]);
    double angulo = atof(argv[4]);
    double direccion = atof(argv[5]);
	float dispersion = atof(argv[6]);;

    distancia = calcularDistanciaProyectil(angulo, velocidadInicial);
    double latitudImpacto, longitudImpacto;
    calcularCoordenadasImpacto(latitudInicial, longitudInicial, direccion, distancia, &latitudImpacto, &longitudImpacto);


 /*
    L2G(latitudImpacto, &gA, &mA, &sA);
    L2G(longitudImpacto, &gO, &mO, &sO);
    printf("impacto:\n%d %d %.2lf ; %d %d %.2lf\n", gA, mA, sA, gO, mO, sO);
    printf("Distancia: %.2lf m\n", distancia);
*/

    // Tu clave de API de Google Maps
    const char* apiKey = "AIzaSyA5FuC-FpZNRVKEl2ZxzEm4DLoC9Mkgg3Y";

    // Definir un vector de coordenadas (latitud, longitud)
    float coordenadas[N+2][2];

	//guardo punto de disparo
	coordenadas[0][0] = latitudInicial;
	coordenadas[0][1] = longitudInicial;

	//printf("calculando impactos");
    // Ingresar las coordenadas
    for (int i = 1; i < N+1; i++) {
		coordenadas[i][0] = addwobble(latitudImpacto,dispersion/10000000);
		coordenadas[i][1] = addwobble(longitudImpacto,dispersion/10000000);
    }

	//guardo centro de impacto
	coordenadas[N+1][0] = latitudImpacto;
	coordenadas[N+1][1] = longitudImpacto;



    // Llamar a la función para generar el código HTML
    generarCodigoHTML(coordenadas, N+2, apiKey);



    return 0;
}

File: web;C/tiro.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
import cgitb, cgi, datetime, sys, re
from reportlab.lib.pagesizes import A4
from reportlab.lib import colors
from reportlab.lib.styles import getSampleStyleSheet
from reportlab.platypus import BaseDocTemplate, SimpleDocTemplate, Paragraph, Table, Frame, PageTemplate,TableStyle, PageBreak
from html import escape
from reportlab.lib.units import inch
from reportlab.lib.units import cm


import random
import math

BOMBAZO="#FF0000"
BOMBAZO_FILL="#8F0000"
POI_A="#FF00FF"
POI_A_FILL="#7F007F"
POI_B="#00FFFF"
POI_B_FILL="#6F6F00"
AMARILLO="#FFFF00"

################################################################################
################################################################################
class GoogleMapsHelper:
    def __init__(self):
        self.objects = []

    def add_circle(self, lat, lng, radius, stroke_color=BOMBAZO, fill_color=BOMBAZO_FILL, stroke_opacity=0.8, fill_opacity=0.35):
        circle = {
            "type": "circle",
            "lat": lat,
            "lng": lng,
            "radius": radius,
            "stroke_color": stroke_color,
            "fill_color": fill_color,
            "stroke_opacity": stroke_opacity,
            "fill_opacity": fill_opacity,
        }
        self.objects.append(circle)

    def add_text_label(self, lat, lng, content):
        text_label = {
            "type": "text_label",
            "lat": lat,
            "lng": lng,
            "content": content,
        }
        self.objects.append(text_label)

    def add_marker(self, lat, lng, valor):
        marker = {
            "type": "marker",
            "lat": lat,
            "lng": lng,
            "label": valor,
        }
        self.objects.append(marker)

    def add_line(self,x0,y0,x1,y1,color,grosor):
        linea = {
            "type": "linea",
            "x0": x0,
            "y0": y0,
            "x1": x1,
            "y1": y1,
            "stroke_color": color,
            "grosor": grosor,
        }
        self.objects.append(linea)

    def add_flecha(self,x0,y0,x1,y1,color=AMARILLO,grosor=5):
        linea = {
            "type": "linea",
            "x0": x0,
            "y0": y0,
            "x1": x1,
            "y1": y1,
            "stroke_color": color,
            "grosor": grosor,
        }
        self.objects.append(linea)

    def add_square(self, lat, lng, size, stroke_color=POI_A, fill_color=POI_B, stroke_opacity=0.8, fill_opacity=0.35):


        square = {
            "type": "square",
            "lat": lat,
            "lng": lng,
            "size": size/55560,
            "stroke_color": stroke_color,
            "fill_color": fill_color,
            "stroke_opacity": stroke_opacity,
            "fill_opacity": fill_opacity,
        }
        self.objects.append(square)


    def generate_js_code(self):
        js_code = []
        for obj in self.objects:
            if obj["type"] == "circle":
                js_code.append(f'new google.maps.Circle({{')
                js_code.append(f'  strokeColor: "{obj["stroke_color"]}",')
                js_code.append(f'  strokeOpacity: {obj["stroke_opacity"]},')
                js_code.append('  strokeWeight: 2,')
                js_code.append(f'  fillColor: "{obj["fill_color"]}",')
                js_code.append(f'  fillOpacity: {obj["fill_opacity"]},')
                js_code.append(f'  map: map,')
                js_code.append(f'  center: {{ lat: {obj["lat"]:.6f}, lng: {obj["lng"]:.6f} }},')
                js_code.append(f'  radius: {obj["radius"]}')
                js_code.append('});')
            elif obj["type"] == "text_label":
                js_code.append(f'var centerLatLng = new google.maps.LatLng({obj["lat"]:.6f}, {obj["lng"]:.6f});')
                js_code.append("var textoEnCentro = new google.maps.InfoWindow({")
                js_code.append(f'  content: "{obj["content"]}",')
                js_code.append("});")
                js_code.append("textoEnCentro.setPosition(centerLatLng);")
                js_code.append("textoEnCentro.open(map);")
            elif obj["type"] == "marker":
                js_code.append(f'new google.maps.Marker({{')
                js_code.append(f' position: {{ lat: {obj["lat"]:.6f}, lng: {obj["lng"]:.6f} }},')
                js_code.append(f'  map: map,')
                js_code.append(f'  label: "{obj["label"]}"')
                js_code.append('});')
            elif obj["type"] == "linea" or obj["type"] == "flecha":
                js_code.append("        var flightPath = new google.maps.Polyline({")
                js_code.append("          path: [")
                js_code.append(f'            {{ lat: {obj["x0"]:.6f}, lng: {obj["y0"]:.6f} }},')
                js_code.append(f'            {{ lat: {obj["x1"]:.6f}, lng: {obj["y1"]:.6f} }},')
                js_code.append("          ],")
                js_code.append("          geodesic: true,")
                if obj["type"] == "flecha":
                    js_code.append("          icons: [{")
                    js_code.append('            icon: { path: google.maps.SymbolPath.FORWARD_CLOSED_ARROW },')
                    js_code.append("            offset: '100%',")
                    js_code.append("          }],")
                js_code.append(f'  strokeColor: "{obj["stroke_color"]}",')
                js_code.append(f'  strokeWeight: {obj["grosor"]},')
                js_code.append("        });")
                js_code.append("        flightPath.setMap(map);")
            elif obj["type"] == "square":
                half_size = obj["size"] / 2
                js_code.append(f'new google.maps.Rectangle({{')
                js_code.append(f'  strokeColor: "{obj["stroke_color"]}",')
                js_code.append(f'  strokeOpacity: {obj["stroke_opacity"]},')
                js_code.append(f'  fillColor: "{obj["fill_color"]}",')
                js_code.append(f'  fillOpacity: {obj["fill_opacity"]},')
                js_code.append(f'  map: map,')
                js_code.append(f'  bounds: {{')
                js_code.append(f'    north: {obj["lat"] + half_size:.6f},')
                js_code.append(f'    south: {obj["lat"] - half_size:.6f},')
                js_code.append(f'    east: {obj["lng"] + half_size:.6f},')
                js_code.append(f'    west: {obj["lng"] - half_size:.6f}')
                js_code.append('  }')
                js_code.append('});')

        return "\n".join(js_code)
################################################################################



cgitb.enable(display=0, logdir='./log')

SYS_NAME='TIRO SERVER'
VERSION='1.1'
DEF_BKGD="#EFEFEF"

form = cgi.FieldStorage()

angulo = float(form.getvalue("elevacion"))
direccion = int(form.getvalue("azimut"))
N = int(form.getvalue("salva"))
dispersion = int(form.getvalue("dispersion")    )
mortero = form.getvalue('mortero')
latitud_inicial = float(form.getvalue('latitud')   )
longitud_inicial = float(form.getvalue('longitud')    )

# Datos de coordenadas tres puntos de interes
punto1 = form.getvalue("punto1")
lat1 = float(form.getvalue('lat1'))
lng1 = float(form.getvalue('lng1'))
punto2 = form.getvalue("punto2")
lat2 = float(form.getvalue('lat2'))
lng2 = float(form.getvalue('lng2'))
punto3 = form.getvalue("punto3")
lat3 = float(form.getvalue('lat3'))
lng3 = float(form.getvalue('lng3'))

################################################################################

def generar_codigo_html(coordenadas, api_key):

    latitudes = [coord[0] for coord in coordenadas]
    longitudes = [coord[1] for coord in coordenadas]
    X = (coordenadas[0][0] + coordenadas[numCoordenadas - 1][0]) / 2;
    Y = (coordenadas[0][1] + coordenadas[numCoordenadas - 1][1]) / 2;

    latitud_media = (max(latitudes) + min(latitudes)) / 2
    longitud_media = (max(longitudes) + min(longitudes)) / 2

    #print("<!DOCTYPE html>")
    print('Content-Type:text/html;charset=utf-8\r\n')
    print('\r\n')
    print("<html>")
    print("  <head>")
    print(f"    <title>Mapa con Coordenadas</title>")
    print(f'    <script src="https://maps.googleapis.com/maps/api/js?key={api_key}"></script>')
    print("  </head>")
    print("  <body>")
    print('    <div id="map" style="height: 800px; width: 100%;"></div>')
    print("    <script>")
    print("      function initMap() {")
    print(f'        var map = new google.maps.Map(document.getElementById("map"), {{')
    print("          zoom: 18,")
    print(f'          center: {{ lat: {latitud_media:.6f}, lng: {longitud_media:.6f} }},')
    print("          mapTypeId: google.maps.MapTypeId.SATELLITE")
    print("        });")

    #creo objeto para agregar items al mapa
    gmaps = GoogleMapsHelper()

	#Recorrer las coordenadas y agregar marcadores
    for i, (lat, lng) in enumerate(coordenadas):

        if i == 0:
            gmaps.add_marker(lat,lng,"Mtr")
        elif i == len(coordenadas) - 1:
            gmaps.add_marker(lat,lng,"Obj")
        else:
            gmaps.add_marker(lat,lng,f'{i}')

        # Agregar el círculo centrado en las mismas coordenadas
        #se saltea el origen y destino
        if i != 0 and i != len(coordenadas)-1:
            gmaps.add_circle(lat,lng,ammoDR)

	#cuadrados en puntos de inrteres
    gmaps.add_square(lat1,lng1,5)
    gmaps.add_square(lat2,lng2,5)
    gmaps.add_square(lat3,lng3,5)

    # Agregar una etiqueta de texto cerca de los POI
    gmaps.add_text_label(lat1,lng1,punto1)
    gmaps.add_text_label(lat2,lng2,punto2)
    gmaps.add_text_label(lat3,lng3,punto3)

	#Agregar una línea con forma de flecha entre el primer y último punto con texto en el centro
    gmaps.add_flecha(coordenadas[0][0],coordenadas[0][1],coordenadas[numCoordenadas - 1][0],coordenadas[numCoordenadas - 1][1])

    #Crear un texto en el centro de la línea
    gmaps.add_text_label(X,Y,f'{distancia:.2f} m, {ammoType}')

    print(gmaps.generate_js_code())

	#cierre del HTML
    print("      }")
    print("    </script>")
    print("    <script>")
    print("      initMap();")
    print("    </script>")
    print("  </body>")
    print("</html>")

################################################################################

def add_wobble(valor_inicial, porcentaje_maximo):
    # Generar un número aleatorio entre -1 y 1
    signo_aleatorio = (random.random() * 2) - 1

    # Calcular el valor aleatorio dentro del rango
    valor_aleatorio = signo_aleatorio * porcentaje_maximo * valor_inicial

    # Calcular el valor final
    valor_final = valor_inicial + valor_aleatorio

    return valor_final
################################################################################

def calcular_distancia_proyectil(angulo, velocidad_inicial):
    gravedad = 9.81

    # Convertir el ángulo de grados a radianes
    angulo_radianes = math.radians(angulo)

    # Calcular la distancia
    distancia = (velocidad_inicial ** 2 * math.sin(2 * angulo_radianes)) / gravedad

    return distancia
################################################################################

def calcular_coordenadas_impacto(latitud_inicial, longitud_inicial, direccion, distancia):
    radio_tierra = 6371000

    # Convertir la dirección en grados a radianes
    direccion_radianes = math.radians(direccion)

    # Convertir latitud y longitud inicial a radianes
    latitud_radianes = math.radians(latitud_inicial)
    longitud_radianes = math.radians(longitud_inicial)

    # Calcular las coordenadas cartesianas del impacto
    x_impacto = distancia * math.cos(direccion_radianes)
    y_impacto = distancia * math.sin(direccion_radianes)

    # Calcular las coordenadas esféricas del impacto
    nueva_latitud_radianes = math.asin(math.sin(latitud_radianes) * math.cos(distancia / radio_tierra) + math.cos(latitud_radianes) * math.sin(distancia / radio_tierra) * math.cos(direccion_radianes))
    nueva_longitud_radianes = longitud_radianes + math.atan2(math.sin(direccion_radianes) * math.sin(distancia / radio_tierra) * math.cos(latitud_radianes), math.cos(distancia / radio_tierra) - math.sin(latitud_radianes) * math.sin(nueva_latitud_radianes))

    # Convertir las coordenadas esféricas del impacto a grados
    latitud_impacto = math.degrees(nueva_latitud_radianes)
    longitud_impacto = math.degrees(nueva_longitud_radianes)

    return latitud_impacto, longitud_impacto
################################################################################

##################### MAIN ######################


#caracteristicas de la municion, velocidad inicial y radio de letalidad
ammoVi = 44
ammoDR = 5
ammoType = 'HE-60mm'

distancia = calcular_distancia_proyectil(angulo, ammoVi)

latitud_impacto, longitud_impacto = calcular_coordenadas_impacto(latitud_inicial, longitud_inicial, direccion, distancia)

coordenadas = [(latitud_inicial, longitud_inicial)]
for _ in range(N):
    latitud = add_wobble(latitud_impacto, dispersion / 10000000)
    longitud = add_wobble(longitud_impacto, dispersion / 10000000)
    coordenadas.append((latitud, longitud))

coordenadas.append((latitud_impacto, longitud_impacto))

numCoordenadas = len(coordenadas)

api_key = "AIzaSyA5FuC-FpZNRVKEl2ZxzEm4DLoC9Mkgg3Y"


generar_codigo_html(coordenadas, api_key)

